### PR TITLE
common: prplmesh_utils.sh: monitor the agents and controller

### PR DESCRIPTION
**Pull request by adam1985d**
On UGW we see multiple beerocks process instances due to 'wait -n'
not working properly and causing prplmesh_utils.sh to exit even if
all beerocks processes are still running.
The 'procd' service, which expects the prplmesh_utils.sh start_wait
command o wait until some failure causes an exit, restarts the
prplmesh_utils.sh which created new instances of the beerocks processes.

Changed the way the start_wait waits for the beerocks_agent and
beerocks_controller processes. If a process crashed, stop_function is
called.

### CODE REVIEW

**Review by vitalybu — APPROVED** [review]

File: common/beerocks/scripts/prplmesh_utils.sh.in [hunk 1]
```diff
@@ -315,7 +316,22 @@ main() {
...
         "start_wait")
             start_function
-            wait -n
+
+            # Store PIDs of the Controller and Agents
+            PREV_PIDS=$(pgrep "beerocks_(agent|controller)")
+            CURR_PIDS=$PREV_PIDS
+            
+            # Monitor the number of running processes
+            while [ ${#CURR_PIDS[@]} -ge ${#PREV_PIDS[@]} ]; do
```
Comment (arnout) [inline common/beerocks/scripts/prplmesh_utils.sh.in]: `${#foo[@]}` is a bashism, and we use /bin/sh. Not OK...

You can just sort the output of pgrep and compare the variables as strings. Or you can wc -l it and compare the integers.

File: common/beerocks/scripts/prplmesh_utils.sh.in [hunk 2]
```diff
@@ -315,7 +316,22 @@ main() {
             ;;
         "start_wait")
             start_function
-            wait -n
+
+            # Store PIDs of the Controller and Agents
+            PREV_PIDS=$(pgrep "beerocks_(agent|controller)")
+            CURR_PIDS=$PREV_PIDS
```
Comment (arnout) [inline common/beerocks/scripts/prplmesh_utils.sh.in]: For consistency with below, I'd write

```
CURR_PIDS="$(pgrep ...)"
PREV_PIDS="$CURR_PIDS"
```

Also the right-hand side really should be quoted, like I did above.

File: common/beerocks/scripts/prplmesh_utils.sh.in [hunk 3]
```diff
@@ -315,7 +316,22 @@ main() {
...
+            while [ ${#CURR_PIDS[@]} -ge ${#PREV_PIDS[@]} ]; do
+                sleep 1
+                CURR_PIDS=$(pgrep "beerocks_(agent|controller)")
+                if [ ${#CURR_PIDS[@]} -gt ${#PREV_PIDS[@]} ]; then
+                        PREV_PIDS=$CURR_PIDS
+                fi
+            done
+            
+            # If one of the main processes stopped, stop the others
+            stop_function
```
Comment (arnout) [inline common/beerocks/scripts/prplmesh_utils.sh.in:341]: Am I craz, or will this *always* kill the just-started processes again?

File: common/beerocks/scripts/prplmesh_utils.sh.in [hunk 4]
```diff
@@ -315,7 +316,22 @@ main() {
...
+            CURR_PIDS=$PREV_PIDS
+            
+            # Monitor the number of running processes
+            while [ ${#CURR_PIDS[@]} -ge ${#PREV_PIDS[@]} ]; do
+                sleep 1
+                CURR_PIDS=$(pgrep "beerocks_(agent|controller)")
+                if [ ${#CURR_PIDS[@]} -gt ${#PREV_PIDS[@]} ]; then
+                        PREV_PIDS=$CURR_PIDS
+                fi
+            done
```
Comment (arnout) [inline common/beerocks/scripts/prplmesh_utils.sh.in]: This entire thing should be moved in a function. The main switch should just call functions that have the real logic.

**Review by arnout — CHANGES_REQUESTED** [review]
Two critical issues:
- bashism;
- kills everything after starting.

I'd like to re-review, so changed requested.

The basic concept LGTM though.

File: common/beerocks/scripts/prplmesh_utils.sh.in [hunk 5]
```diff
@@ -131,6 +131,7 @@ prplmesh_agent_start() {
 prplmesh_agent_stop() {
     echo "prplmesh_agent_stop - stopping beerocks_agent process..."
     killall_program beerocks_agent
+    killall_program beerocks_monitor
```
Comment (tomereli) [inline common/beerocks/scripts/prplmesh_utils.sh.in:134]: This should be in a separate commit.

**Review by dob71 — APPROVED** [review]

**Review by tomereli — APPROVED** [review]
@arnout 's comments should be addressed, other than those I think it is good to go.

File: common/beerocks/scripts/prplmesh_utils.sh.in (hunk 1 above)
Comment (vitalybu) [reply common/beerocks/scripts/prplmesh_utils.sh.in]: Bashism or not, if it works in the ash shell as well, why not keep it? :)

File: common/beerocks/scripts/prplmesh_utils.sh.in (hunk 3 above)
Comment (vitalybu) [reply common/beerocks/scripts/prplmesh_utils.sh.in:341]: The while loop will break only if one of the monitored processes exists.
In this case, we want to cleanup everything.
Perhaps this case can be renamed to "run_wait"

File: common/beerocks/scripts/prplmesh_utils.sh.in (hunk 2 above)
Comment (vitalybu) [reply common/beerocks/scripts/prplmesh_utils.sh.in]: What are the benefits of quoting in this case?

Comment (rmelotte) [reply common/beerocks/scripts/prplmesh_utils.sh.in]: Not quoting the right hand side of an assignment is also a bashism. With some shells, you'll get CURR_PIDS being set to the first pid, then the shell will probably try to execute the next one as if it was a command.

Comment (vitalybu) [reply common/beerocks/scripts/prplmesh_utils.sh.in]: Fixed

File: common/beerocks/scripts/prplmesh_utils.sh.in (hunk 4 above)
Comment (vitalybu) [reply common/beerocks/scripts/prplmesh_utils.sh.in]: Moved

File: common/beerocks/scripts/prplmesh_utils.sh.in (hunk 1 above)
Comment (rmelotte) [reply common/beerocks/scripts/prplmesh_utils.sh.in]: I agree with @arnout.
Since this particular script can potentially be used on many different platforms, you can't rely on the fact that some specific implementations do support it.

File: common/beerocks/scripts/prplmesh_utils.sh.in [hunk 6]
```diff
@@ -141,6 +142,27 @@ prplmesh_delete_logs() {
     rm -rf /tmp/${SUDO_USER:-${USER}}/beerocks/logs
 }
 
+prplmesh_watchdog() {
+    # Store PIDs of the Controller and Agents
+    local CURR_PIDS="$(pgrep 'beerocks_(agent|controller)')"
+    local PREV_PIDS="$CURR_PIDS"
+
```
Comment (rmelotte) [inline common/beerocks/scripts/prplmesh_utils.sh.in]: There are some extra spaces here (nitpick)

File: common/beerocks/scripts/prplmesh_utils.sh.in (hunk 1 above)
Comment (vitalybu) [reply common/beerocks/scripts/prplmesh_utils.sh.in]: Same assumption is true to many other parts of the this script.
Is it safe to assume that all platform support `pgrep` or `wc`?

Comment (rmelotte) [reply common/beerocks/scripts/prplmesh_utils.sh.in]: True, except that you can make it a platform requirement to have, say `procps` available.
You can also make it a requirement to have bash installed for instance (although we don't want to impose that), but it doesn't mean that `/bin/sh` will run bash.
Since `/bin/sh` could really be any shell, the general common denominator we can assume is that it's going to be a POSIX-compliant shell.

The problem is not about the availability of the tools being used, but rather about using features available across different implementations of the same tool (the shell in this case). In the cases where we can, why not do it?

Comment (vitalybu) [reply common/beerocks/scripts/prplmesh_utils.sh.in]: No reason, I was just curious about where we'd like to draw the compatibility line.
Pushed a fixup with all the changes.

File: common/beerocks/scripts/prplmesh_utils.sh.in (hunk 6 above)
Comment (vitalybu) [reply common/beerocks/scripts/prplmesh_utils.sh.in]: Removed

**Review by tomereli — APPROVED** [review]

**Review by rmelotte — COMMENTED** [review]
The new solution looks far from ideal. There should definitely be a better solution than running a few processes every second to check if everything is still running.
Why did `wait -n` 'not work' in the beginning? What specific exit code does it exit with?
Is it because the `-n` option is not available?

How about something like this:
```
start_wait() {
    run "$1"
    kill the main script
}
# main script
trap 'kill $(jobs -p)' EXIT
start_wait agent &
start_wait controller &
wait
```

Would that work in this case?

If `wait` (without any option) works as it should, this should also exit prplmesh_utils.sh as soon as a child exits.

Maybe someone else will have another idea on how to do it.

File: common/beerocks/scripts/prplmesh_utils.sh.in (hunk 3 above)
Comment (adam1985d) [reply common/beerocks/scripts/prplmesh_utils.sh.in:341]: > The while loop will break only if one of the monitored processes exists.
to clarify - @vitalybu meant "exits" instead of "exists"

File: common/beerocks/scripts/prplmesh_utils.sh.in (hunk 5 above)
Comment (adam1985d) [reply common/beerocks/scripts/prplmesh_utils.sh.in:134]: done

**Comment by vitalybu** [comment]
> The new solution looks far from ideal. There should definitely be a better solution than running a few processes every second to check if everything is still running.
> Why did `wait -n` 'not work' in the beginning? What specific exit code does it exit with?
> Is it because the `-n` option is not available?
> 
> How about something like this:
> 
> ```
> start_wait() {
>     run "$1"
>     wait
>     kill the main script
> }
> # main script
> trap 'kill $(jobs -p)' EXIT
> start_wait agent &
> start_wait controller &
> wait
> ```
> 
> Would that work in this case?
> 
> If `wait` (without any option) works as it should, this should also exit prplmesh_utils.sh as soon as a child exits.
> 
> Maybe someone else will have another idea on how to do it.

It's an intermediate solution, until we'll create a real daemon process to start everything.
The busybox on UGW doesn't have the `-n` option for `wait`.
`wait` without any arguments waits for all child processes to die before returning, and that's not the behavior we want to achieve.

**Comment by tomereli** [comment]
I verified this on UGW and it indeed solves the issue. It may not be ideal, but it works.
@vitalybu - can we start working on a daemon process which will start everything instead of all these hacks? How long do you estimate something like this should take?

**Comment by vitalybu** [comment]
> I verified this on UGW and it indeed solves the issue. It may not be ideal, but it works.
> @vitalybu - can we start working on a daemon process which will start everything instead of all these hacks? How long do you estimate something like this should take?

I'll take this task as soon as I'm back from Ischgl (WW9).

**Comment by rmelotte** [comment]
> It's an intermediate solution, until we'll create a real daemon process to start everything.
> The busybox on UGW doesn't have the `-n` option for `wait`.
> `wait` without any arguments waits for all child processes to die before returning, and that's not the behavior we want to achieve.

That's why I suggested killing the parent from all the children when they stop, you don't need `wait -n` (note that there was a wait too much in start_wait). The main script either waits for all the children, or gets killed by the first one to finish. It's also a hack anyway that's for sure...

File: common/beerocks/scripts/prplmesh_utils.sh.in (hunk 3 above)
Comment (arnout) [reply common/beerocks/scripts/prplmesh_utils.sh.in:341]: Right, now I get it, thanks!

Did you change the commit message? Because now it is clear to me and yesterday it wasn't, so *something* has changed :-)

**Review by arnout — APPROVED** [review]
This PR has become *way* simpler and clearer than it originally was. Hurray for reviews!
